### PR TITLE
[multicast_dns] Allow decode of binary encoded TXT

### PR DESCRIPTION
**Pull request by ajhemphill91**
Follow up on https://github.com/flutter/flutter/issues/141218. Added a unit test with a binary-encoded payload.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I [linked to at least one issue that this PR fixes] in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

**Comment by ajhemphill91** [comment]
Side note, the decoding algorithm for TXT data is still not exactly in spec (see [RFC6763 6.3-6.5](https://datatracker.ietf.org/doc/html/rfc6763#section-6.3)). 

In particular:

> DNS-SD TXT record strings
   beginning with an '=' character (i.e., the key is missing) MUST be
   silently ignored.

> The characters of a key MUST be printable US-ASCII values (0x20-0x7E)
   [[RFC20](https://datatracker.ietf.org/doc/html/rfc20)], excluding '=' (0x3D).

These don't matter to me too much at the present time. It does look like the openthread border router's [implementation](https://github.com/openthread/ot-br-posix/blob/897028fe9ef1b7604d8911f1c31b048d70c15c8a/src/mdns/mdns.cpp#L132) has these restrictions on the '=' character within the key.

**Comment by stuartmorgan** [comment]
From triage: ping @dnfield on this review

File: packages/multicast_dns/lib/src/resource_record.dart
Comment (jmagman) [inline packages/multicast_dns/lib/src/resource_record.dart:341]: Apologies for the delay in this review.

I believe this would be a breaking change to [TxtResourceRecord](https://pub.dev/documentation/multicast_dns/latest/multicast_dns/TxtResourceRecord/TxtResourceRecord.html)?

How about a new named constructor:

```dart
  TxtResourceRecord.attributes(
    String name,
    int validUntil, {
    required List<String> attributes,
  }) : this(name, validUntil, text: attributes.join('\n'));
```
Also total nit but `attributes` instead of `attrs`?

**Comment by jmagman** [comment]
Hi @ajhemphill91, have you been able to circle back to the requested change for the named constructor?  Appreciate your contribution!

**Comment by ajhemphill91** [comment]
Sorry for the delay, I have been on another project recently. If you do have a preference to keep the original constructor (and thus the original ```final String text``` property) for API stability, your suggested changes sound good. Internally, this constructor is only referenced once in packet.dart and then in the unit tests which are updated by this commit. 

The ```text``` property as a String is not valid since it represents a list of [encoded strings](http://www.zeroconf.org/Rendezvous/txtrecords.html) which [are treated as binary information](https://datatracker.ietf.org/doc/html/rfc1035#section-3.3). Basically it mutates the record by adding the ```\n``` characters.

Actually now that I am looking at it, ```encodeResponseRecord``` ought to fail to produce correct results too since the record is mutated by the ```text``` property. However it looks as though this property is never referenced anywhere.

**Comment by stuartmorgan** [comment]
From triage: @ajhemphill91 Are you still planning on updating this PR?

**Comment by ajhemphill91** [comment]
I can apply the nit, but changing back to a named constructor that joins the records with a newline nullifies this PR in the first place. What would you like me to do?

**Comment by jmagman** [comment]
Sorry, didn't realize this PR was waiting for my response.

> but changing back to a named constructor that joins the records with a newline nullifies this PR in the first place


Hm, yeah I can't think of a way to do this without something unpleasant like a `text.split('\n')`
```dart
  TxtResourceRecord(
    String name,
    int validUntil, {
    required this.text,
  })  : attrs = text.split('\n'),
        super(ResourceRecordType.text, name, validUntil);

  TxtResourceRecord.attributes(
    String name,
    int validUntil, {
    required this.attrs,
  })  : text = attrs.join('\n'),
        super(ResourceRecordType.text, name, validUntil);
```

Adding @cbracken for his opinion here.  Ideally this could be done without a breaking change.

**Comment by stuartmorgan** [comment]
@ajhemphill91 I've re-added the checklist from the template to the PR description; in the future please don't remove it. This PR is missing several required steps (notably for this discussion, a version change).

**Comment by ajhemphill91** [comment]
Per the [versioning guide](https://dart.dev/tools/pub/versioning) should I then update the pubspec.yaml version to 0.4.0? I can take care of that and the CHANGELOG.md if everyone is okay with the breaking change.

**Comment by stuartmorgan** [comment]
@jmagman Are you okay with the breaking change?

**Comment by jmagman** [comment]
> @jmagman Are you okay with the breaking change?

LGTM, I can't remember if this will impact the tool but it would be easy to fix up

File: packages/multicast_dns/lib/src/resource_record.dart
Comment (cbracken) [reply packages/multicast_dns/lib/src/resource_record.dart:341]: A way to do this without it being a breaking change would be to:
1. Make `text` nullable and mark it deprecated.
2. Add `attrs` as nullable.
3. Throw if (a) both or (b) neither are null.

After the deprecation period, mark `attrs` as non-nullable and remove `text`.

If very few packages are relying on this, then it might not be worth the effort.

**Review by cbracken — APPROVED** [review]
This seems fine to me. Suggestion on a non-breaking way to do this in my comment above.

**Comment by cbracken** [comment]
@ajhemphill91 Looks like Dart formatting and CHANGELOG format checks are failing. See the log here for details:
https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20repo_checks/4671/overview

Once those are done, I leave it up to you as to whether you want to do the multi-step dance to go through the deprecation, then breaking change removal, or just go straight to the breaking change and land this as-is.

**Comment by stuartmorgan** [comment]
@ajhemphill91 Is this still something you are planning on updating?

**Comment by stuartmorgan** [comment]
Thank you for your contribution. I'm going to close this PR for now since there are outstanding comments, just to get this off our PR review queue. Please don't hesitate to submit a new PR if you have the time to address the review comments. Thanks!

**Pull request closed without merging**
